asf_heat_pump_suitability/pipeline/run_scripts/run_add_features.py

```diff
@@ -139,28 +144,23 @@ def parse_arguments() -> argparse.Namespace:
     logging.info("Adding property density to EPC")
     epc_df = property_density.extend_df_with_property_density(epc_df)
 
+    # Add feature: property density Scotland
+    dz_density_df = property_density.generate_df_property_density_s()
+    epc_df = epc_df.join(dz_density_df, how="left", left_on="lsoa", right_on="DataZone")
+
     # Add feature: off gas postcodes
     logging.info("Adding off gas grid column to EPC")
     off_gas_postcodes = off_gas.process_off_gas_data()
     epc_df = off_gas.add_off_gas_feature(epc_df, off_gas_postcodes)
 
     # Add feature: listed buildings data
-    logging.info("Loading listed buildings for England")
-    e_listed_buildings_df = listed_buildings.transform_gdf_listed_buildings("England")
-    e_listed_buildings_df = listed_buildings.sjoin_df_epc_with_listed_buildings(
-        epc_df, e_listed_buildings_df
+    logging.info("Adding listed buildings to EPC")
+    listed_buildings_df = listed_buildings.generate_df_epc_listed_buildings(
+        epc_df=epc_df
     )
-
-    logging.info("Loading listed buildings for Wales")
-    w_listed_buildings_df = listed_buildings.transform_gdf_listed_buildings("Wales")
-    w_listed_buildings_df = listed_buildings.sjoin_df_epc_with_listed_buildings(
-        epc_df, w_listed_buildings_df
-    )
-
-    listed_buildings_df = pl.concat(
-        [e_listed_buildings_df, w_listed_buildings_df], how="vertical"
+    epc_df = epc_df.join(listed_buildings_df, how="left", on="UPRN").with_columns(
+        pl.col("listed_building").fill_null(False)
```

Comment:
I'm not sure where it happened but I also have two columns `UPRN` and `UPRN_right`. It seems that all the english data has null for `UPRN`.

Reply:
TLDR: this will be fixed in a separate PR because it originates in the reweighting script not the add features script.

Thanks for noticing this! I looked into it. It seems to come from the reweighting pipeline because it already exists in the dataset that we load in at the beginning of `run_add_features.py`. I think it originates from the [outer join](https://github.com/nestauk/asf_heat_pump_suitability/blob/a75c867fe0b4ef85182e4d90292a3cb1edeb1693/asf_heat_pump_suitability/pipeline/run_scripts/run_compute_epc_weights.py#L151) we use to join the weights to the `epc_df`. This is an expected `polars` behaviour that I forgot to account for. 

This means, rows which have nulls in `UPRN_right` should be missing weights. They will either be missing weights because:

- They are in an LSOA which was skipped because we are missing census data for it (mainly affects Scotland, but also England and Wales in some instances)
- We had to drop the row for weighting because the row had a category which was not found in the target data for that LSOA. E.g. if the property type of the row is 'flat' but the census has 0% flats for that LSOA, the row will be dropped.
- They are in Scotland. This dataset was run before reweighting for Scotland was added, so all Scottish rows should be missing a weight.

Here is the count of rows with missing `UPRN_right` for each country:
```
>>> epc.filter(pl.col("UPRN_right").is_null())["COUNTRY"].value_counts()
shape: (3, 2)
┌──────────┬─────────┐
│ COUNTRY  ┆ count   │
│ ---      ┆ ---     │
│ cat      ┆ u32     │
╞══════════╪═════════╡
│ Wales    ┆ 38285   │
│ England  ┆ 788475  │
│ Scotland ┆ 1473612 │
```

All of them have null weights, as expected:
```
>>> epc.filter(pl.col("UPRN_right").is_null())["weight"].value_counts()
shape: (1, 2)
┌────────┬─────────┐
│ weight ┆ count   │
│ ---    ┆ ---     │
│ f64    ┆ u32     │
╞════════╪═════════╡
│ null   ┆ 2300372 │
```